### PR TITLE
[TBD] G Suite Admin Integration

### PR DESCRIPTION
**Pull request by crestdatasystems**
## Contributing to Cortex XSOAR Content
Make sure to register your contribution by filling the [contribution registration form](https://forms.gle/XDfxU4E61ZwEESSMA)

**The Pull Request will be reviewed only after the contribution registration form is filled.**

## Status
- [ ] In Progress
- [x] Ready
- [ ] In Hold - (Reason for hold)

## Description
G Suite or Google Workspace Admin is an integration to perform an action on IT infrastructure, create users, update settings, and more administrative tasks. 

## Minimum version of Demisto
- [x] 5.0.0
- [ ] 5.5.0
- [ ] 6.0.0

## Does it break backward compatibility?
   - [ ] Yes
       - Further details:
   - [x] No

## Must have
- [x] Tests
- [x] Documentation

### CODE REVIEW

**Comment by orhovy** [comment]
@crestdatasystems Please fill the [contribution form](https://forms.gle/XDfxU4E61ZwEESSMA) so we can start the review.

**Comment by crestdatasystems** [comment]
> @crestdatasystems Please fill the [contribution form](https://forms.gle/XDfxU4E61ZwEESSMA) so we can start the review.

@orhovy we have filled contribution form. Thanks

File: Packs/ApiModules/Scripts/GSuiteApiModule/GSuiteApiModule.py [hunk 1]
```diff
@@ -0,0 +1,306 @@
...
+        except ValueError:
+            message = COMMON_MESSAGES['UNEXPECTED_ERROR']
+
+        if response[0].status in status_code_message_map:
+            raise DemistoException(status_code_message_map[response[0].status].format(message))
+        else:
+            raise DemistoException(COMMON_MESSAGES['UNKNOWN_ERROR'].format(response[0].status, message))
+
+    @staticmethod
+    def safe_load_non_strict_json(json_string: str) -> Dict[str, Any]:
```
Comment (orhovy) [inline Packs/ApiModules/Scripts/GSuiteApiModule/GSuiteApiModule.py:207]: Similar functionality exists in `commonServerPython` in a function `safe_load_json`. 
Please us it if possible.

File: Packs/ApiModules/Scripts/GSuiteApiModule/GSuiteApiModule.py [hunk 2]
```diff
@@ -0,0 +1,306 @@
...
+        """
+        try:
+            if json_string:
+                return json.loads(json_string, strict=False)
+            return {}
+        except ValueError:
+            raise ValueError(COMMON_MESSAGES['JSON_PARSE_ERROR'])
+
+    @staticmethod
+    def strip_dict(args: Dict[str, str]) -> Dict[str, str]:
```
Comment (orhovy) [inline Packs/ApiModules/Scripts/GSuiteApiModule/GSuiteApiModule.py:224]: Utility function should be outside of client class

File: Packs/ApiModules/Scripts/GSuiteApiModule/GSuiteApiModule.py [hunk 3]
```diff
@@ -0,0 +1,306 @@
...
+        Converts all dictionary keys from snake case (dict_key) to lower camel case(dictKey).
+        :param dictionary: Dictionary which may contain keys in snake_case
+        :return: Dictionary with snake_case keys converted to lowerCamelCase
+        """
+        underscore_pattern = re.compile(r'_([a-z])')
+        return {underscore_pattern.sub(lambda i: i.group(1).upper(), key.lower()): value for (key, value) in
+                dictionary.items()}
+
+    @staticmethod
+    def validate_get_int(max_results: Optional[str], message: str, limit: int = 0) -> Optional[int]:
```
Comment (orhovy) [inline Packs/ApiModules/Scripts/GSuiteApiModule/GSuiteApiModule.py:285]: Utility function should be outside of client class

**Review by orhovy — COMMENTED** [review]
Looks great, just need to change some small changes.
- Some utility function of apimodule already exist in CommonServerPython, its better practice to use the existing functions instead of duplicating code.
- The utility function in the apimodule should be outside of the client, this will cause some changes around the code.
- Make sure that context entries in the integration which are set to `Unknown` could not be set to a specific type.

File: Packs/GSuiteAdmin/Integrations/GSuiteAdmin/GSuiteAdmin.py [hunk 4]
```diff
@@ -0,0 +1,1063 @@
...
+''' COMMAND FUNCTIONS '''
+
+
+@logger
+def test_function(client) -> str:
+    """
+    Performs test connectivity by valid http response
+
+    :param client: client object which is used to get response from api.
+    :param last_run: Demisto last run dictionary.
```
Comment (orhovy) [inline Packs/GSuiteAdmin/Integrations/GSuiteAdmin/GSuiteAdmin.py:471]: Please remove this docstring as its not a param of the function

File: Packs/GSuiteAdmin/Integrations/GSuiteAdmin/GSuiteAdmin.py [hunk 5]
```diff
@@ -0,0 +1,1063 @@
...
+        elif command in commands:
+            return_results(commands[command](gsuite_client, args))
+
+        # Log exceptions
+    except Exception as e:
+        demisto.error(traceback.format_exc())
+        return_error(f'Error: {str(e)}')
+
+
+from GSuiteApiModule import *  # noqa: E402
```
[the remark this answers is not in the document]
Comment (crestdatasystems) [reply Packs/GSuiteAdmin/Integrations/GSuiteAdmin/GSuiteAdmin.py:1060]: @orhovy we took a ref of [MicrosoftApiModule](https://github.com/demisto/content/blob/master/Packs/ApiModules/Scripts/MicrosoftApiModule/README.md)

Let us know if there is any convention change or any document.

File: Packs/GSuiteAdmin/Integrations/GSuiteAdmin/GSuiteAdmin.py [hunk 6]
```diff
@@ -0,0 +1,1063 @@
...
+    """
+    Takes an action that affects a mobile device. For example, remotely wiping a device.
+
+    :param client: client object used to get response from api
+    :param args: command arguments
+    :return: CommandResults which returns detailed results to war room and sets the context data.
+    """
+
+    client.set_authorized_http(scopes=COMMAND_SCOPES['MOBILE_UPDATE'], subject=ADMIN_EMAIL)
+    args.pop('admin_email', '')
```
[the remark this answers is not in the document]
Comment (crestdatasystems) [reply Packs/GSuiteAdmin/Integrations/GSuiteAdmin/GSuiteAdmin.py:498]: @orhovy our purpose is not to send admin_email in the request body of mobile update API.

File: Packs/ApiModules/Scripts/GSuiteApiModule/GSuiteApiModule.py [hunk 7]
```diff
@@ -0,0 +1,306 @@
...
+        if arg in args:
+            if args[arg].lower() == 'true':
+                args[arg] = True
+            elif args[arg].lower() == 'false':
+                args[arg] = False
+            else:
+                raise ValueError(COMMON_MESSAGES['BOOLEAN_ERROR'].format(arg_name if arg_name else arg))
+
+    @staticmethod
+    def remove_empty_entities(d):
```
[the remark this answers is not in the document]
Comment (crestdatasystems) [reply Packs/ApiModules/Scripts/GSuiteApiModule/GSuiteApiModule.py:254]: @orhovy As you are not removing empty values ("") we decided to have this method with the addition of this feature.
So in our case, we want to remove both null and empty values.

File: Packs/ApiModules/Scripts/GSuiteApiModule/GSuiteApiModule.py (hunk 1 above)
Comment (crestdatasystems) [reply Packs/ApiModules/Scripts/GSuiteApiModule/GSuiteApiModule.py:207]: @orhovy this method is basically used when someone copy and paste JSON secret key which contains /n char. Due to this, safe_load_json throws error so we used not only for test_module function but for other JSON safe loading.

File: Packs/GSuiteAdmin/pack_metadata.json [hunk 8]
```diff
@@ -0,0 +1,17 @@
+{
+    "name": "G Suite Admin",
+    "description": "G Suite Admin integration with Cortex XSOAR. G Suite or Google Workspace Admin is an integration to perform an action on IT infrastructure, create users, update settings, and more administrative tasks.",
+    "support": "partner",
+    "currentVersion": "1.0.0",
+    "author": "G Suite Admin",
+    "url": "https://www.paloaltonetworks.com/cortex",
+    "email": "",
```
[the remark this answers is not in the document]
Comment (crestdatasystems) [reply Packs/GSuiteAdmin/pack_metadata.json:8]: @orhovy this integration was requested by the PaloAltoNetwork team, can you please suggest us for the same? We feel that it should be the sales email i.e: contact_sales@paloaltonetworks.com. 
Please confirm.

[the remark this answers is not in the document]
Comment (anara123) [reply Packs/GSuiteAdmin/pack_metadata.json:8]: @orhovy @crestdatasystems these integrations should be supported by `xsoar`
```
    "support": "xsoar",
    "currentVersion": "1.0.0",
    "author": "Cortex XSOAR",
    "url": "https://www.paloaltonetworks.com/cortex",
    "email": "",
```

**Comment by orhovy** [comment]
@crestdatasystems  Regarding the documentation, the integration readme file should contain step by step explanation regarding how to create your service account and what exact roles should be added for each command.
Please take a look at [GoogleKubernetesEngine](https://github.com/demisto/content/tree/master/Packs/GoogleKubernetesEngine/Integrations/GoogleKubernetesEngine) readme file for reference. 
As every user should be able to set a service account and configure the integration using our readme.

File: Packs/ApiModules/Scripts/GSuiteApiModule/test_data/service_account_json.txt [hunk 9]
```diff
@@ -0,0 +1,12 @@
+{
```
[the remark this answers is not in the document]
Comment (crestdatasystems) [reply Packs/ApiModules/Scripts/GSuiteApiModule/test_data/service_account_json.txt:1]: It's just a dummy format of a service account JSON.

**Review by orhovy — APPROVED** [review]

**Pull request merged**